aoc/pkg/advent/advent.go

```diff
@@ -0,0 +1,108 @@
+package advent
+
+import (
+	"aoc/pkg/structs"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+	"os"
+	"strings"
+)
+
+func buildClient(urlObj *url.URL) (http.Client, error) {
+
+	sessionId := os.Getenv("AOC_SESSION_ID")
+	if sessionId == "" {
+		return http.Client{}, errors.New("no AOC_SESSION_ID provided. " +
+			"(It can be found by logging in through a browser and getting the session cookie.)",
+		)
+	}
+	jar, err := cookiejar.New(nil)
+	if err != nil {
+		return http.Client{}, err
+	}
+	client := http.Client{
+		Jar: jar,
+	}
+	sessionCookie := &http.Cookie{
+		Name:  "session",
+		Value: os.Getenv("AOC_SESSION_ID"),
+	}
+	client.Jar.SetCookies(urlObj, []*http.Cookie{sessionCookie})
+	return client, nil
+}
+
+func GetInputs(day int, year int) ([]structs.Group, error) {
+
+	urlObj, _ := url.Parse(fmt.Sprintf("https://adventofcode.com/%d/day/%d/input", year, day))
+	client, err := buildClient(urlObj)
+	if err != nil {
+		return make([]structs.Group, 0), err
```

Comment:
Unless you specifically need an empty slice when error occurs, you should return the zero value of slice which is `nil`

Reply:
My mental model was something like "if its an error who cares about the value anyways" and so I just wrote whatever made the compiler happy. 

In practise, does the `Ok` value really matter if the `err` returned is non-nil?

Reply:
It's more idiomatic to return nil slice with error. In contrast to nil slice, the 0-length make the call to make for allocation, which you don't need here.

Sometimes you do want a 0-length slice specifically. For example, if you want to convert to json string, a `nil` slice converts to `null` and a 0-length slice converts to `[]`

Apart from idiomatic reasons it doesn't matter in most cases when returned with non-nil error.

Reply:
gotcha! `nil` ftw